src/flag_gems/ops/nllloss.py

```diff
@@ -0,0 +1,283 @@
+    tl.store(inp_grad_ptrs, inp_grad.to(tl.float32), mask=(mask_inp & ignore_mask))
+
+
+@libentry()
+@triton.autotune(
+    configs=[
+        triton.Config({"BLOCK_C": c, "BLOCK_D": d}, num_warps=4)
+        for c in [256, 512, 1024]
+        for d in [1, 4, 16]
+    ],
+    key=["C", "D"],
+)
+@triton.jit(do_not_specialize=["ignore_index"])
+def nll_loss_multi_fwd_kernel(
+    inp_ptr,
+    tgt_ptr,
+    w_ptr,
+    w_tgt_ptr,
+    out_ptr,
+    ignore_index,
+    N,
+    C,
+    D,
+    BLOCK_C: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    pid_d = tl.program_id(1)
+    offset_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
+
+    tgt_ptrs = tgt_ptr + pid_n * D + offset_d
+    mask_tgt = offset_d < D
+    tgt = tl.load(tgt_ptrs, mask=mask_tgt, other=0)
+
+    ignore_mask = not (tgt == ignore_index)
+
+    w_ptrs = w_ptr + tgt
+    w_tgt = tl.load(w_ptrs, mask=mask_tgt, other=0).to(tl.float32)
+    w_tgt_ptrs = w_tgt_ptr + pid_n * D + offset_d
+    tl.store(w_tgt_ptrs, w_tgt, mask=(mask_tgt & ignore_mask))
+
+    inp_tgt_ptrs = inp_ptr + pid_n * C * D + tgt * D + offset_d
+    inp_tgt = tl.load(inp_tgt_ptrs, mask=mask_tgt, other=-float("inf")).to(tl.float32)
+    out = inp_tgt * w_tgt * -1
+    out_ptrs = out_ptr + pid_n * D + offset_d
+    tl.store(out_ptrs, out, mask=(mask_tgt & ignore_mask))
+
+
+@libentry()
+@triton.autotune(
+    configs=[
+        triton.Config({"BLOCK_C": c, "BLOCK_D": d}, num_warps=4)
+        for c in [256, 512, 1024]
+        for d in [1, 4, 16]
+    ],
+    key=["C", "D"],
+)
+@triton.jit(do_not_specialize=["ignore_index", "total_weight"])
+def nll_loss_multi_bwd_kernel(
+    out_grad_ptr,
+    tgt_ptr,
+    w_ptr,
+    inp_grad_ptr,
+    ignore_index,
+    total_weight,
+    N,
+    C,
+    D,
+    BLOCK_C: tl.constexpr,
+    BLOCK_D: tl.constexpr,
+):
+    pid_n = tl.program_id(0)
+    pid_d = tl.program_id(1)
+    offset_d = pid_d * BLOCK_D + tl.arange(0, BLOCK_D)
+
+    tgt_ptrs = tgt_ptr + pid_n * D + offset_d
+    mask_tgt = offset_d < D
+    tgt = tl.load(tgt_ptrs, mask=mask_tgt, other=0)
+
+    ignore_mask = (tgt != ignore_index)[None, :]
+
+    w_ptrs = w_ptr + tgt
+    w_tgt = tl.load(w_ptrs, mask=mask_tgt, other=0).to(tl.float32)[None, :]
+    out_grad_ptrs = out_grad_ptr + pid_n * D + offset_d
+    out_grad = (tl.load(out_grad_ptrs, mask=mask_tgt, other=0).to(tl.float32))[None, :]
+
+    for off in range(0, C, BLOCK_C):
+        offset_c = off + tl.arange(0, BLOCK_C)
+        inp_mask = offset_c[:, None] < C and offset_d[None, :] < D
+        inp_mask = inp_mask and offset_c[:, None] == tgt
+        inp_grad = -1 * out_grad * w_tgt / total_weight
+        inp_grad_ptrs = (
+            inp_grad_ptr + pid_n * C * D + offset_c[:, None] * D + offset_d[None, :]
+        )
+        tl.store(inp_grad_ptrs, inp_grad.to(tl.float32), mask=(inp_mask & ignore_mask))
+
+
+class NLLLoss(torch.autograd.Function):
```

Comment:
This function is intended be used as substitute for `nll_loss` whereas NLLLoss is already taken as the nn module name. We should avoid the name confusion.

Reply:
Indeed. I've upated the class name.